// sphere-node-sync/lib/utils/product-utils.js
var ProductUtils, Utils, actionsList, buildAddPriceAction, buildNewSetAttributeAction, buildRemovePriceAction, buildSetAttributeAction, helper, jsondiffpatch, _, _ref,
  __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

_ = require("underscore")._;

jsondiffpatch = require("jsondiffpatch");

Utils = require("./utils");

helper = require("../helper");

/*
Product Utils class
*/


ProductUtils = (function(_super) {
  var actionsMapVariantAttributes;

  __extends(ProductUtils, _super);

  function ProductUtils() {
    _ref = ProductUtils.__super__.constructor.apply(this, arguments);
    return _ref;
  }

  ProductUtils.prototype.diff = function(old_obj, new_obj) {
    var patchEnum, patchEnums, patchPrices;
    patchPrices = function(obj) {
      if (obj.masterVariant) {
        if (obj.masterVariant.prices && obj.masterVariant.prices.length > 0) {
          _.each(obj.masterVariant.prices, function(p, i) {
            return p._id = i;
          });
        }
      }
      if (obj.variants && obj.variants.length > 0) {
        return _.each(obj.variants, function(v) {
          if (v.prices && v.prices.length > 0) {
            return _.each(v.prices, function(p, i) {
              return p._id = i;
            });
          }
        });
      }
    };
    patchPrices(old_obj);
    patchPrices(new_obj);
    patchEnum = function(attribute) {
      var v;
      if (_.has(attribute.value, 'key') && _.has(attribute.value, 'label')) {
        v = attribute.value.key;
        delete attribute.value;
        return attribute.value = v;
      }
    };
    patchEnums = function(obj) {
      if (obj.masterVariant) {
        if (obj.masterVariant.attributes) {
          _.each(obj.masterVariant.attributes, function(attrib, index) {
            return patchEnum(attrib);
          });
        }
      }
      if (obj.variants) {
        return _.each(obj.variants, function(variant) {
          if (variant.attributes) {
            return _.each(variant.attributes, function(attrib, index) {
              return patchEnum(attrib);
            });
          }
        });
      }
    };
    patchEnums(old_obj);
    patchEnums(new_obj);
    return ProductUtils.__super__.diff.call(this, old_obj, new_obj);
  };

  ProductUtils.prototype.actionsMap = function(diff, old_obj) {
    var actions;
    actions = [];
    _.each(actionsList(), function(item) {
      var a, action, key, keys, obj, old, updated;
      key = item.key;
      action = (function() {
        switch (key) {
          case "name":
          case "slug":
          case "description":
            obj = diff[key];
            if (obj) {
              updated = {};
              if (_.isArray(obj)) {
                updated = helper.getDeltaValue(obj);
              } else {
                keys = _.keys(obj);
                _.each(keys, function(k) {
                  var value;
                  value = helper.getDeltaValue(obj[k]);
                  return updated[k] = value;
                });
              }
              old = _.clone(old_obj[key]);
              _.extend(old, updated);
              a = {
                action: item.action
              };
              if (updated) {
                a[key] = old;
              } else {
                a[key] = void 0;
              }
              return a;
            }
        }
      })();
      if (action) {
        return actions.push(action);
      }
    });
    return actions;
  };

  ProductUtils.prototype.actionsMapPrices = function(diff, old_obj, new_obj) {
    var actions, prices;
    actions = [];
    if (diff.masterVariant) {
      prices = diff.masterVariant.prices;
      if (prices) {
        _.each(prices, function(value, key) {
          var addAction, index, removeAction;
          if (key.match(/^\d$/g)) {
            index = key;
          } else if (key.match(/^\_\d$/g)) {
            index = key.substring(1);
          }
          if (index) {
            removeAction = buildRemovePriceAction(old_obj.masterVariant, index);
            if (removeAction) {
              actions.push(removeAction);
            }
            addAction = buildAddPriceAction(new_obj.masterVariant, index);
            if (addAction) {
              return actions.push(addAction);
            }
          }
        });
      }
    }
    if (diff.variants) {
      _.each(diff.variants, function(variant, i) {
        prices = variant.prices;
        if (prices) {
          return _.each(prices, function(value, key) {
            var addAction, index, removeAction;
            if (key.match(/^\d$/g)) {
              index = key;
            } else if (key.match(/^\_\d$/g)) {
              index = key.substring(1);
            }
            if (index) {
              removeAction = buildRemovePriceAction(old_obj.variants[i], index);
              if (removeAction) {
                actions.push(removeAction);
              }
              addAction = buildAddPriceAction(new_obj.variants[i], index);
              if (addAction) {
                return actions.push(addAction);
              }
            }
          });
        }
      });
    }
    return _.sortBy(actions, function(a) {
      return a.action === "addPrice";
    });
  };

  actionsMapVariantAttributes = function(attributes, variant, sameForAllAttributeNames) {
    var actions;
    actions = [];
    if (attributes) {
      _.each(attributes, function(value, key) {
        var id, index, setAction, v;
        if (key.match(/^\d$/g)) {
          if (_.isArray(value)) {
            v = helper.getDeltaValue(value);
            id = variant.id;
            setAction = buildNewSetAttributeAction(id, v);
            if (setAction) {
              return actions.push(setAction);
            }
          } else {
            index = key;
            setAction = buildSetAttributeAction(value.value, variant, index, sameForAllAttributeNames);
            if (setAction) {
              return actions.push(setAction);
            }
          }
        } else if (key.match(/^\_\d$/g)) {
          if (_.isArray(value)) {
            v = helper.getDeltaValue(value);
            if (!v) {
              v = value[0];
              delete v.value;
            }
            id = variant.id;
            setAction = buildNewSetAttributeAction(id, v);
            if (setAction) {
              return actions.push(setAction);
            }
          } else {
            index = key.substring(1);
            setAction = buildSetAttributeAction(value.value, variant, index, sameForAllAttributeNames);
            if (setAction) {
              return actions.push(setAction);
            }
          }
        }
      });
    }
    return actions;
  };

  ProductUtils.prototype.actionsMapAttributes = function(diff, new_obj, sameForAllAttributeNames) {
    var actions, attributes, mActions, masterVariant;
    if (sameForAllAttributeNames == null) {
      sameForAllAttributeNames = [];
    }
    actions = [];
    masterVariant = diff.masterVariant;
    if (masterVariant) {
      attributes = masterVariant.attributes;
      mActions = actionsMapVariantAttributes(attributes, new_obj.masterVariant, sameForAllAttributeNames);
      actions = actions.concat(mActions);
    }
    if (diff.variants) {
      _.each(diff.variants, function(variant, i) {
        var vActions;
        attributes = variant.attributes;
        vActions = actionsMapVariantAttributes(attributes, new_obj.variants[i], sameForAllAttributeNames);
        return actions = actions.concat(vActions);
      });
    }
    return _.unique(actions, function(action) {
      return JSON.stringify(action);
    });
  };

  return ProductUtils;

})(Utils);

/*
Exports object
*/


module.exports = ProductUtils;

actionsList = function() {
  return [
    {
      action: "changeName",
      key: "name"
    }, {
      action: "changeSlug",
      key: "slug"
    }, {
      action: "setDescription",
      key: "description"
    }
  ];
};

buildRemovePriceAction = function(variant, index) {
  var action, price;
  price = variant.prices[index];
  if (price) {
    delete price._id;
    action = {
      action: "removePrice",
      variantId: variant.id,
      price: price
    };
  }
  return action;
};

buildAddPriceAction = function(variant, index) {
  var action, price;
  price = variant.prices[index];
  if (price) {
    delete price._id;
    action = {
      action: "addPrice",
      variantId: variant.id,
      price: price
    };
  }
  return action;
};

buildSetAttributeAction = function(diffed_value, variant, index, sameForAllAttributeNames) {
  var action, attribute, centAmount, currencyCode, text;
  attribute = variant.attributes[index];
  if (attribute) {
    action = {
      action: "setAttribute",
      variantId: variant.id,
      name: attribute.name
    };
    if (_.contains(sameForAllAttributeNames, attribute.name)) {
      action.action = 'setAttributeInAllVariants';
      delete action.variantId;
    }
    if (_.isArray(diffed_value)) {
      action.value = helper.getDeltaValue(diffed_value);
    } else {
      if (_.isString(diffed_value)) {
        action.value = helper.getDeltaValue(diffed_value);
      } else if (diffed_value.centAmount) {
        if (diffed_value.centAmount) {
          centAmount = helper.getDeltaValue(diffed_value.centAmount);
        } else {
          centAmount = attribute.value.centAmount;
        }
        if (diffed_value.currencyCode) {
          currencyCode = helper.getDeltaValue(diffed_value.currencyCode);
        } else {
          currencyCode = attribute.value.currencyCode;
        }
        action.value = {
          centAmount: centAmount,
          currencyCode: currencyCode
        };
      } else if (_.isObject(diffed_value)) {
        text = {};
        _.each(diffed_value, function(v, k) {
          return text[k] = helper.getDeltaValue(v);
        });
        action.value = text;
      }
    }
  }
  return action;
};

buildNewSetAttributeAction = function(id, el, sameForAllAttributeNames) {
  var action, attributeName;
  attributeName = el.name;
  action = {
    action: "setAttribute",
    variantId: id,
    name: attributeName,
    value: el.value
  };
  if (_.contains(sameForAllAttributeNames, attributeName)) {
    action.action = 'setAttributeInAllVariants';
    delete action.variantId;
  }
  return action;
};
